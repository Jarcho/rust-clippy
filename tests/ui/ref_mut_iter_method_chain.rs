// run-rustfix
#![warn(clippy::ref_mut_iter_method_chain)]

macro_rules! m {
    ($i:ident) => {
        $i
    };
    (&mut $i:ident) => {
        &mut $i
    };
    (($i:expr).$m:ident($arg:expr)) => {
        ($i).$m($arg)
    };
}

fn main() {
    let mut iter = [0, 1, 2].iter();
    let _ = (&mut iter).find(|&&x| x == 1);
    let _ = (&mut m!(iter)).find(|&&x| x == 1);

    // Don't lint. `&mut` comes from macro expansion.
    let _ = m!(&mut iter).find(|&&x| x == 1);

    // Don't lint. Method call from expansion
    let _ = m!((&mut iter).find(|&&x| x == 1));

    // Don't lint. No method chain.
    for &x in &mut iter {
        print!("{}", x)
    }

    let iter = &mut iter;
    (&mut *iter).find(|&&x| x == 1);
}
