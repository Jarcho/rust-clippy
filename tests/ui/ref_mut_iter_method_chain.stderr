error: use of `&mut` on an iterator in a method chain
  --> $DIR/ref_mut_iter_method_chain.rs:18:13
   |
LL |     let _ = (&mut iter).find(|&&x| x == 1);
   |             ^^^^^^^^^^^ help: try: `iter.by_ref()`
   |
   = note: `-D clippy::ref-mut-iter-method-chain` implied by `-D warnings`

error: use of `&mut` on an iterator in a method chain
  --> $DIR/ref_mut_iter_method_chain.rs:19:13
   |
LL |     let _ = (&mut m!(iter)).find(|&&x| x == 1);
   |             ^^^^^^^^^^^^^^^ help: try: `m!(iter).by_ref()`

error: use of `&mut` on an iterator in a method chain
  --> $DIR/ref_mut_iter_method_chain.rs:32:15
   |
LL |     for &x in (&mut iter).filter(|&&x| x % 2 == 0) {
   |               ^^^^^^^^^^^ help: try: `iter.by_ref()`

error: use of `&mut` on an iterator in a method chain
  --> $DIR/ref_mut_iter_method_chain.rs:37:5
   |
LL |     (&mut *iter).find(|&&x| x == 1);
   |     ^^^^^^^^^^^^ help: try: `iter.by_ref()`

error: use of `&mut` on an iterator in a method chain
  --> $DIR/ref_mut_iter_method_chain.rs:40:5
   |
LL |     (&mut **iter).find(|&&x| x == 1);
   |     ^^^^^^^^^^^^^ help: try: `(*iter).by_ref()`

error: aborting due to 5 previous errors

